### PR TITLE
Allow both CPU and GPU training

### PR DESCRIPTION
**Pull request by hsheil**

### CODE REVIEW

**Comment by hsheil** [comment]
Sample output when training on the CPU..

yeats:lstm hsheil$ th main.lua 
Could not find cunn or fbcunn. Falling back to CPU  
Loading ./data/ptb.train.txt, size of data = 929589 
Loading ./data/ptb.valid.txt, size of data = 73760  
Loading ./data/ptb.test.txt, size of data = 82430  
Network parameters: 
{
  max_grad_norm : 5
  seq_length : 20
  batch_size : 20
  lr : 1
  max_max_epoch : 13
  rnn_size : 200
  init_weight : 0.1
  decay : 2
  dropout : 0
  layers : 2
  vocab_size : 10000
  max_epoch : 4
}
Creating a RNN LSTM network.  
Starting training.  
epoch = 0.004, train perp. = 9951.720, wps = 951, dw:norm() = 4.807, lr = 1.000, since beginning = 0 mins.  
epoch = 0.104, train perp. = 7745.948, wps = 961, dw:norm() = 2.678, lr = 1.000, since beginning = 2 mins.  
epoch = 0.204, train perp. = 5717.303, wps = 975, dw:norm() = 3.366, lr = 1.000, since beginning = 3 mins.  
epoch = 0.304, train perp. = 4085.200, wps = 980, dw:norm() = 4.418, lr = 1.000, since beginning = 5 mins.  
epoch = 0.404, train perp. = 2866.210, wps = 983, dw:norm() = 4.500, lr = 1.000, since beginning = 6 mins.  
epoch = 0.504, train perp. = 1990.892, wps = 985, dw:norm() = 4.819, lr = 1.000, since beginning = 8 mins.  
epoch = 0.604, train perp. = 1356.728, wps = 985, dw:norm() = 4.618, lr = 1.000, since beginning = 9 mins.  
epoch = 0.703, train perp. = 921.981, wps = 980, dw:norm() = 4.133, lr = 1.000, since beginning = 11 mins.  
epoch = 0.803, train perp. = 622.625, wps = 981, dw:norm() = 4.612, lr = 1.000, since beginning = 13 mins.  
epoch = 0.903, train perp. = 415.353, wps = 982, dw:norm() = 4.678, lr = 1.000, since beginning = 14 mins.  
Validation set perplexity : 182.621

**Comment by soumith** [comment]
there is a branch called "cpu" https://github.com/wojzaremba/lstm/tree/cpu
It achieves this. 
Since CPU is a use-case that is not really an option for us (because of speed), we'd rather not have the code have additional clutter for this case.

**Comment by hsheil** [comment]
Hi @soumith - ok understood. My motivation was to provide CPU compatibility for when the user doesn't have access to GPUs / quick, local prototyping and especially if this code became the basis for LSTM / RNN in Torch - but it looks like the rnn library from Nicholas Leonard will become that..
